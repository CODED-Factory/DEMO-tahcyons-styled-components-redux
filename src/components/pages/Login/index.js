import React, { useReducer, useEffect } from "react";
import { connect } from "react-redux";

// Components
import TextFieldGroup from "../../common/TextFieldGroup";
import Alert from "../../common/Alert";

// Styles
import LoginStyles from "./styles/LoginStyles";

// Local Reducers
import userDataReducer from "./userDataReducer";

// Actions
import { login, resetErrors } from "../../../redux/actions";

export const Login = ({ user, errors, login, resetErrors }) => {
  const initialState = {
    username: "",
    password: ""
  };

  const [{ username, password }, dispatch] = useReducer(
    userDataReducer,
    initialState
  );

  useEffect(() => resetErrors, [resetErrors]);

  const onChange = e =>
    dispatch({ type: e.target.name, payload: e.target.value });

  const onSubmit = e => {
    e.preventDefault();
    login({ username, password });
  };

  return (
    <LoginStyles className="pa6 black-80">
      <form
        className="measure center bg-white pa4 br2"
        onSubmit={onSubmit}
        data-test="login-form"
      >
        <fieldset id="login" className="ba b--transparent ph0 mh0">
          <legend className="f4 fw6 ph0 mh0">Sign In</legend>
          {errors.detail && <Alert level="error" text={errors.detail} />}
          <TextFieldGroup
            name="username"
            value={username}
            label="Username"
            placeholder="Mr Potato"
            info="a username do you have one?"
            error={errors.username}
            onChange={onChange}
            required
          />
          <TextFieldGroup
            name="password"
            type="password"
            value={password}
            label="Password"
            info="a password do you have one?"
            error={errors.password}
            onChange={onChange}
            required
          />
          <input
            className="b ph3 pv2 input-reset ba b--black grow pointer f6 dib"
            type="submit"
          />
        </fieldset>
      </form>
    </LoginStyles>
  );
};

const mapStateToProps = ({ errors, auth }) => ({
  user: auth.user,
  errors
});

const mapDispatchToProps = {
  login,
  resetErrors
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);
