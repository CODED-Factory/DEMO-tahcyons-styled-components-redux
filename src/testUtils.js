import casual from "casual";

casual.seed(666);

export const fakeUserData = overrides => ({
  username: casual.username,
  password: casual.password,
  ...overrides
});

export const type = (wrapper, name, value) => {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
};
