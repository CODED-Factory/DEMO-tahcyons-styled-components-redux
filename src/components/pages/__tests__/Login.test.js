import React from "react";
import { shallow, mount } from "enzyme";

// Components
import { Login } from "../Login";

// Test Utils
import { type, fakeUserData } from "../../../testUtils";

describe("<Login />", () => {
  const mockLogin = jest.fn();
  const userData = fakeUserData();
  const { username, password } = userData;

  it("renders and matches snapshot", () => {
    const wrapper = shallow(<Login errors={{}} />);
    expect(wrapper.find('[data-test="login-form"]')).toMatchSnapshot();
  });

  it("updates the form contents when there are changes", () => {
    const wrapper = mount(<Login errors={{}} />);
    type(wrapper, "username", username);
    type(wrapper, "password", password);
    expect(wrapper.find('[data-test="login-form"]')).toMatchSnapshot();
  });

  it("renders errors on the form", () => {
    const wrapper = shallow(
      <Login
        errors={{
          username: ["this field cannot be blank"],
          password: ["this field cannot be blank"]
        }}
      />
    );
    expect(wrapper.find('[data-test="login-form"]')).toMatchSnapshot();
  });

  it("submits the user data", async () => {
    const wrapper = mount(<Login login={mockLogin} errors={{}} />);
    type(wrapper, "username", username);
    type(wrapper, "password", password);

    wrapper.find("form").simulate("submit");

    expect(mockLogin).toHaveBeenCalledTimes(1);
    expect(mockLogin).toHaveBeenCalledWith(userData);
  });
});
