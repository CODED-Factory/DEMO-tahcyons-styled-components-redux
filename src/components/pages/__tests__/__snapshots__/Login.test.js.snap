// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders and matches snapshot 1`] = `
<form
  className="measure center bg-white pa4 br2"
  data-test="login-form"
  onSubmit={[Function]}
>
  <fieldset
    className="ba b--transparent ph0 mh0"
    id="login"
  >
    <legend
      className="f4 fw6 ph0 mh0"
    >
      Sign In
    </legend>
    <TextFieldGroup
      info="a username do you have one?"
      label="Username"
      name="username"
      onChange={[Function]}
      placeholder="Mr Potato"
      required={true}
      type="text"
      value=""
    />
    <TextFieldGroup
      info="a password do you have one?"
      label="Password"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
    <input
      className="b ph3 pv2 input-reset ba b--black grow pointer f6 dib"
      type="submit"
    />
  </fieldset>
</form>
`;

exports[`<Login /> renders errors on the form 1`] = `
<form
  className="measure center bg-white pa4 br2"
  data-test="login-form"
  onSubmit={[Function]}
>
  <fieldset
    className="ba b--transparent ph0 mh0"
    id="login"
  >
    <legend
      className="f4 fw6 ph0 mh0"
    >
      Sign In
    </legend>
    <TextFieldGroup
      error={
        Array [
          "this field cannot be blank",
        ]
      }
      info="a username do you have one?"
      label="Username"
      name="username"
      onChange={[Function]}
      placeholder="Mr Potato"
      required={true}
      type="text"
      value=""
    />
    <TextFieldGroup
      error={
        Array [
          "this field cannot be blank",
        ]
      }
      info="a password do you have one?"
      label="Password"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
    <input
      className="b ph3 pv2 input-reset ba b--black grow pointer f6 dib"
      type="submit"
    />
  </fieldset>
</form>
`;

exports[`<Login /> updates the form contents when there are changes 1`] = `
<form
  className="measure center bg-white pa4 br2"
  data-test="login-form"
  onSubmit={[Function]}
>
  <fieldset
    className="ba b--transparent ph0 mh0"
    id="login"
  >
    <legend
      className="f4 fw6 ph0 mh0"
    >
      Sign In
    </legend>
    <TextFieldGroup
      info="a username do you have one?"
      label="Username"
      name="username"
      onChange={[Function]}
      placeholder="Mr Potato"
      required={true}
      type="text"
      value="Barton_Sporer"
    >
      <div
        className="measure mt3"
      >
        <label
          className="db fw6 lh-copy f6"
          htmlFor="username"
        >
          Username
          <span
            className="normal red"
          >
             *
          </span>
        </label>
        <input
          aria-describedby="name-desc"
          className="pa2 input-reset ba hover-bg-black hover-white w-100  b--black-20"
          id="username"
          name="username"
          onChange={[Function]}
          placeholder="Mr Potato"
          type="text"
          value="Barton_Sporer"
        />
        <small
          className="f6 black-60 db mb2"
          id="name-desc"
        >
          a username do you have one?
        </small>
      </div>
    </TextFieldGroup>
    <TextFieldGroup
      info="a password do you have one?"
      label="Password"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value="7Cory98"
    >
      <div
        className="measure mt3"
      >
        <label
          className="db fw6 lh-copy f6"
          htmlFor="password"
        >
          Password
          <span
            className="normal red"
          >
             *
          </span>
        </label>
        <input
          aria-describedby="name-desc"
          className="pa2 input-reset ba hover-bg-black hover-white w-100  b--black-20"
          id="password"
          name="password"
          onChange={[Function]}
          type="password"
          value="7Cory98"
        />
        <small
          className="f6 black-60 db mb2"
          id="name-desc"
        >
          a password do you have one?
        </small>
      </div>
    </TextFieldGroup>
    <input
      className="b ph3 pv2 input-reset ba b--black grow pointer f6 dib"
      type="submit"
    />
  </fieldset>
</form>
`;
